pkg/env/env.go

```diff
@@ -1,13 +1,29 @@
-// Package for processing environment variables.
 package env
 
-// TODO: we need to add tests for this package.
-
 import (
-	"bufio"
 	"fmt"
+	"io"
 	"os"
 	"strings"
+
+	"github.com/containers/storage/pkg/regexp"
+)
+
+var (
+	// Form: https://github.com/motdotla/dotenv/blob/aa03dcad1002027390dac1e8d96ac236274de354/lib/main.js#L9C76-L9C76
+	// ([\w.-]+) match key
+	// ([\w.-]+)(\s*[=|*]\s*?|:\s+?) match separator
+	// Remaining match value
+	// e.g. KEY=VALUE => KEY, =, VALUE
+	//
+	//	KEY= => KEY, =, ""
+	//	KEY* => KEY, *, ""
+	//	KEY*=1 => KEY, *, =1
+	lineRegexp = regexp.Delayed(
+		`(?m)(?:^|^)\s*(?:export\s+)?([\w.%-]+)(\s*[=|*]\s*?|:\s+?)(\s*'(?:\\'|[^'])*'|\s*"(?:\\"|[^"])*"|\s*` +
+			"`(?:\\`|[^`])*`" + `|[^#\r\n]+)?\s*(?:#.*)?(?:$|$)`,
+	)
```

Comment:
Did you consider using an external library for parsing?

I wonder whether we can use github.com/hashicorp/go-envparse and add our logic of "*" on top.  I don't know whether that will work but I would like to try it out.

The regex above is hard to maintain.

Reply:
TL;DR: `go-envparse` is not suitable for our needs.

I have previously looked into the `go-envparse` library, and it does not support the following scenarios:
1. It cannot handle `ENV*` prefix matching.
2. It ignores whitespace in value assignments, such as `K= V`.

Unless we fork it or patch it to support our requirements.

Currently, the regex we are using is based on https://github.com/motdotla/dotenv, with only minimal modifications.

Reply:
> Unless we fork it or patch it to support our requirements.

Forking is probably not a nice way to approach it.  In a way, we already fork parts of it (see regex) which is why I am worried.  Did you consider opening an issue at `go-envparse`?

Reply:
This may not be a good idea because our requirements are not 100% compatible with `dotenv`. Therefore, submitting issues or PRs to `go-envparse` is likely to be closed.

Reply:
We only know once we tried.  I think it's good practice to avoid fragmentation.  On the long run, it will be less work on the Podman as the code is maintained outside of Podman.

Reply:
@Luap99 WDYT?

Reply:
We can ask but I am not sure why they would want to support our awkward syntax. The readme clearly states:
> The following common features are intentionally missing:
>    ...

I would love to just use it as is but we must stay backwards compatible at least until 5.0. 

But I also do not really like this regex, I cannot understand it. I tried for far to long and just gave up. I know this is not your fault, regex syntax is just bad. And I know a handwritten parser is likely not any better.

Reply:
Thanks! So let's continue with this PR as is?

Reply:
> But I also do not really like this regex, I cannot understand it. I tried for far to long and just gave up. I know this is not your fault, regex syntax is just bad. And I know a handwritten parser is likely not any better.

I spent several hours understanding the principle of this regular expression while writing it.
Currently, we can validate whether this regular expression is backward compatible through a large number of unit tests, just as written on the side of this PR.

Reply:
Yes tests are great, as of today we have only a single basically no tests for --env-file so this is welcome.
But I really cannot comment on the regex and have no time to understand it.

I fear that we could be missing corner cases and might break existing people but well I cannot offer anything useful either so I ok with it.

What I would like is an explicit documentation on what the support syntax is in the man page. Currently all we have is `Read in a line-delimited file of environment variables.` which is not helpful to users trying to figure out how to use it.

Reply:
> What I would like is an explicit documentation on what the support syntax is in the man page. Currently all we have is Read in a line-delimited file of environment variables. which is not helpful to users trying to figure out how to use it.

Good idea! I think I can modify the description for `--env-file` like the following.
To prevent triggering multiple CI runs, we can check if the following description is acceptable. If there are no issues, I will submit a commit to update the content of `--env-file` description.

```md
Read the environment variables from the file, supporting prefix matching: `KEY*`, as well as multiline values in double quotes and single quotes, but not multiline values in backticks.
The env-file will ignore comments, empty lines, and spaces or tabs before and after the KEY.
If an invalid value is encountered, such as only an `=` sign, it will be skipped.
If it is a prefix match (`KEY*`), all environment variables starting with KEY on the host machine will be loaded.
If it is only KEY (`KEY`), the KEY environment variable on the host machine will be loaded.
To be compatible with Bash syntax, the `--env-file` will also support the `%%=` syntax, such as: `BASH_FUNC__echo%%=() { echo 1; }`.
```

Reply:
@BlackHole1 I don't see a man-page change in the PR yet.  Could you add that before merging?

Reply:
Of course, but it's the weekend now, so I will submit it on next Monday.

Reply:
Absolutely!  Enjoy the weekend :+1:

Reply:
@vrothberg Done 😃.